File: leetcode.c
/***********************************************************
** 2018-01-17
** Liu Chunlin / chunlins@outlook.com
** Solution for problem on leetcode.com.
** License on MIT
**
** FORMATER
** No. $problem_number | $problem_title | $difficulty
** $description
** $solution_one
** $code_one
** ...
** $testcase
**
** CONTENTS
**  * Problems
**  * Collections
**  * Solutions
************************************************************/


/***********************************************************
** Problems
** NO. 001 | Two Sum | Easy
************************************************************/

/***********************************************************
** TOP INTERVIEW QUESTIONS **
** https://leetcode.com/explore/interview/card/top-interview-questions-easy/
** 
************************************************************/

/*
** No. 001 | Two Sum | Easy
** Given an array of integers, return indices of the two numbers such that 
** they add up to a specific target.
** You may assume that each input would have exactly one solution, and you
** may not use the same element twice.
**
** Example:
**  Given nums = [2, 7, 11, 15], target = 9,
**  Because nums[0] + nums[1] = 2 + 7 = 9,
**  return [0, 1].
*/

// Note: The returned array must be malloced, assume caller calls free().
int* twoSum(int* nums, int numsSize, int target) {
    int i, j;
        int indices[] = {0, 0};
        for (i = 0; i < (numsSize - 1); i++)
            for (j = i + 1; j < numsSize; j++)
                if ( nums[i] + nums[j] == target)
                {
                    indices[0] = i;
                    indices[1] = j;
                    return indices;
                }
        return indices;
}


/*
** No. XXX | Rotate Array |
** Rotate an array of n elements to the right by k steps.
** For example, with n = 7 and k = 3, the array [1,2,3,4,5,6,7] is 
** rotated to [5,6,7,1,2,3,4].
**
** Note:
** Try to come up as many solutions as you can, there are at least 
** 3 different ways to 
** solve this problem.
** 
** Hint:
** Could you do it in-place with O(1) extra space?
** Related problem: Reverse Words in a String II
** 
** Credits:
** Special thanks to @Freezen for adding this problem and creating all test cases.
*/

void rotate(int* nums, int numsSize, int k) {
    int temps[numsSize];
    int i;
    for (i = 0; i < numsSize; i += 1)
        temps[(i + k) % numsSize] = nums[i];
    for (i = 0; i < numsSize; i += 1)
        nums[i] = temps[i];
}




/* 
** No. 136 | Single Number | Easy
** Given an array of integers, every element appears twice except for one.
** Find that single one.
**
** Note:
** Your algorithm should have a linear runtime complexity. 
** Could you implement it without using extra memory? 
*/

/* NORMAL SOLUTION */
int 
singleNumberSelf(int* nums, int numsSize) {
    int i, j, k;
    for (i = 0, j = numsSize-1; i <= numsSize / 2; i ++, j--)
        for (k = j; k > i; k--)
            if (nums[k] == nums[j])
            {
                // exchange nums[j] and nums[k]
                nums[k] = nums[k] + nums[j];
                nums[j] = nums[k] - nums[j];
                nums[k] = nums[k] - nums[j];
            }
    return nums[numsSize / 2];
}

/* USE XOR OPERATION */
int
singleNumberXOR(int* nums, int numsSize) {
    int i, result;
    for (i = 0; i < numsSize; i++)
        result ^= nums[numsSize];
    return result;
}

void rotate(int** matrix, int matrixRowSize) {
    int next = 0;
    int length, next, row, col;
    for (length = matrixRowSize; length > 1; length -= 2) {
        row = 0; col = 0;
        for (itera = row; itera < length - 1; itera++) {
            // matrix[row][col + itera]
            // matrix[col + itera][length - 1]
            // matrix[lenght - 1][lenght - 1 - col - itera]
            // matrix[col + itera][col]
            next = matrix[col + itera][length - 1];
            matrix[col + itera][length - 1] = matrix[row][col + itera];
            next = matrix[lenght - 1][lenght - 1 - col - itera];
            matrix[lenght - 1][lenght - 1 - col - itera] = matrix[col + itera][length - 1];
            next = matrix[col + itera][col]
            matrix[col + itera][col] = matrix[lenght - 1][lenght - 1 - col - itera];
            next = matrix[row][col + itera];
            matrix[row][col + itera] = matrix[col + itera][col];
        }
        row++; col++;
    }
}

void main(void) {
    int matrix[2][2] = {{1, 2}, {3, 4}};
    rotate(matrix, 2);
    for (int i = 0; i < 2; i++) { 
        for (int j = 0; j < 2; j++) 
            printf("%3d,");
        printf("\n");
    }
}

File: leetcode.py
"""
Solutions for problem on leetcode.com

Author: Liu Chunlin (chll) | <chunlins@outlook.com>
Created: 2018-01-17, modified at 2018-04-03
License: MIT LICENSE

FORMATER:
    No. $problem_number | $problem_title | $difficulty
    $description
    $solution_one
    $code_one
    ...
    $testcase

CONTENTS:
    * Problems
    * Collections
    * Solutions
"""
""" 
Problems
    NO. 001 | Two Sum | Easy
    No. 020 | Valid Parenthese | Easy
"""
"""
** TOP INTERVIEW QUESTIONS **
https://leetcode.com/explore/interview/card/top-interview-questions-easy/
"""


"""
No. 11 | Container With Most Water | Media

Given n non-negative integers a1, a2, ..., an, where each represents a point
at coordinate (i, ai). n vertical lines are drawn such that the two endpoints
of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis
forms a container, such that the container contains the most water.

Note: You may not slant the container and n is at least 2.
"""
# MySolution, iterate over start and end pointer
def maxArea(self, height):
    count, area = len(height), []
    for start in range(0, count):
        for end in range(start, count):
            hightof = min(height[start], height[end])
            lengthof = end - start
            area.append(hightof * lengthof)
    
    return max(area)

# MySolution, iterate over axis length
def maxArea(self, height):
    area = []
    for axis in range(1, len(height)):
        water = [min(height[i], height[i - axis]) 
                 for i in range(axis, len(height))]
        area.append(max(water) * axis)
        
    return max(area)


"""
No. 020 | Valid Parenthese | Easy

Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
determine if the input string is valid.

The brackets must close in the correct order, "()" and "()[]{}" are all valid
but "(]" and "([)]" are not.

Input:
  '(', '{}', '{}', '{()]}, '{()}()]]'
"""

def isValid(self, s):
    valid = True
    stack = []
    brackets = {'(': ')', '[': ']', '{': '}'}

    if len(s) % 2 == 1:
        valid = False

    for ch in s:
        if ch in brackets.keys():
            stack.append(ch)
        else:
            if len(stack) == 0:
                valid = False
                continue
            check = stack.pop()
            if check != brackets.get(ch):
                valid = False

    return valid


"""
No. 412 | Fizz Buzz | Easy

Write a program that outputs the string representation of numbers from 1 to n.

But for multiples of three it should output “Fizz” instead of the number and for the multiples of five output “Buzz”. For numbers which are multiples of both three and five output “FizzBuzz”.

Example:
    n = 6,
    Return: ["1", "2", "Fizz", "4", "Buzz", "Fizz"]
"""

def fizzBuzz(n):
    """ n: int / return: List[str] """
    if n <= 0:
        return []
    
    result = [str(number) for number in range(n + 1)]
    for index in range(0, n + 1, 3):
        result[index] = 'Fizz'
    for index in range(0, n + 1, 5):
        result[index] = 'Buzz'
    for index in range(0, n + 1, 15):
        result[index] = 'FizzBuzz'
    
    return result[1:]
